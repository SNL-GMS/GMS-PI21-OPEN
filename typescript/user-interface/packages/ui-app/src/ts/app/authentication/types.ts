import type { AuthenticationStatus } from '@gms/ui-state';

export interface Authenticator {
  authenticateWith(userName: string): Promise<AuthenticationStatus>;
  checkIsAuthenticated(): Promise<AuthenticationStatus>;
  unAuthenticateWith(): Promise<AuthenticationStatus>;
  logout(setAppAuthenticationStatus: (status: AuthenticationStatus) => void): void;
}

export interface LoginRequestParams {
  userName: string;
}
