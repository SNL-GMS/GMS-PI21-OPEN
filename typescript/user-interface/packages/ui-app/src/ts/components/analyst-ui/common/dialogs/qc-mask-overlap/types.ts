import type { QcMaskTypes } from '@gms/common-model';

import type { CoordinatesPx } from '../qc-mask-form/types';
import type { QcMaskDialogBoxType } from '../types';

/**
 * QcMaskOverlap Props
 */
export interface QcMaskOverlapProps {
  masks: QcMaskTypes.QcMask[];
  // If set, any context menus opened as children will be created at given x/y coordinates
  contextMenuCoordinates?: CoordinatesPx;
  openNewContextMenu(
    x: number,
    y: number,
    mask: QcMaskTypes.QcMask,
    dialogType: QcMaskDialogBoxType
  );
  selectMask(mask: QcMaskTypes.QcMask);
}

/**
 * QcMaskOverlap  State
 */
export interface QcMaskOverlapState {
  selectedMask: QcMaskTypes.QcMask;
  createDialogOfType?: QcMaskDialogBoxType;
}
export interface QcMaskTableButtonParams {
  onClick(x: number, y: number, params: any);
}
