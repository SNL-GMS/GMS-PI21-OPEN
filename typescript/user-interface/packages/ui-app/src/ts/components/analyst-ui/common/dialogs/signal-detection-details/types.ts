import type { SignalDetectionTypes } from '@gms/common-model';
import type { Row } from '@gms/ui-core-components';

/**
 * SignalDetectionDetails Props
 */
export interface SignalDetectionDetailsProps {
  detection: SignalDetectionTypes.SignalDetection;
  color: string;
  assocStatus?: string;
}

/**
 * SignalDetectionDetails State
 */
export interface SignalDetectionDetailsState {
  showHistory: boolean;
}

/**
 * Interface that describes the Detection history
 * information.
 */
export interface SignalDetectionHistoryRow extends Row {
  id: string;
  versionId: string;
  creationTimestamp: number;
  phase: string;
  arrivalTimeMeasurementFeatureType: string;
  arrivalTimeMeasurementTimestamp: number;
  arrivalTimeMeasurementUncertaintySec: number;
  channelSegmentType: string;
  authorName: string;
  rejected: string;
}
