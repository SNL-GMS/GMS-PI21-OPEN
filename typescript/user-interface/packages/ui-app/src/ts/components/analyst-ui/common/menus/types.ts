export interface HideStationMenuItemProps {
  stationName: string;
  hideStationCallback: (stationName: any) => void;
  showHideText?: string;
  disabled?: boolean;
}
