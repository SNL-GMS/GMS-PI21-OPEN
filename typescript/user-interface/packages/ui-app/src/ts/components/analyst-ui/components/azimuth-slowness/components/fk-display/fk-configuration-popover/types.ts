import type { FkTypes } from '@gms/common-model';
import type { CheckboxSearchListTypes } from '@gms/ui-core-components';

import type {
  FkConfigurationWithUnits,
  FkUnits
} from '~analyst-ui/components/azimuth-slowness/types';

export enum FkConfigurationPopoverPanel {
  DEFAULT,
  ADVANCED
}

export interface FkConfigurationPopoverState {
  openPanel: FkConfigurationPopoverPanel;
  fkUnits: FkUnits;
  normalizeWaveforms: boolean;
  mediumVelocity: number;
  maximumSlowness: number;
  numberOfPoints: number;
  channelCheckboxes: CheckboxSearchListTypes.CheckboxItem[];
  useVerticalChannelOffsets: boolean;
}
export interface FkConfigurationPopoverProps extends FkTypes.FkConfiguration {
  fkUnitDisplayed: FkUnits;
  applyFkConfiguration(configuration: FkConfigurationWithUnits);
  close();
}
