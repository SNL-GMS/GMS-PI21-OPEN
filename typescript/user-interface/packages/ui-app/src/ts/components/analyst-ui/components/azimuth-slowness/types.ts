import type {
  CommonTypes,
  FkTypes,
  LegacyEventTypes,
  ProcessingStationTypes,
  SignalDetectionTypes,
  WaveformTypes,
  WorkflowTypes
} from '@gms/common-model';
import type GoldenLayout from '@gms/golden-layout';
import type {
  AnalystWorkspaceTypes,
  ProcessingAnalystConfigurationQueryProps
} from '@gms/ui-state';
import type Immutable from 'immutable';

import type {
  FilterType,
  FkThumbnailSize
} from './components/fk-thumbnail-list/fk-thumbnails-controls';
import type { LeadLagPairs } from './constants';

export interface FkParams {
  windowParams: FkTypes.WindowParameters;
  frequencyPair: FkTypes.FrequencyBand;
}

export interface LeadLagPairAndString {
  leadLagPairs: LeadLagPairs;
  windowParams: FkTypes.WindowParameters;
}

export enum FkUnits {
  FSTAT = 'FSTAT',
  POWER = 'POWER'
}

/**
 * Used to return a super set of the fk configuration from the fk config popover
 */
export interface FkConfigurationWithUnits extends FkTypes.FkConfiguration {
  fkUnitToDisplay: FkUnits;
}

/**
 * Azimuth Slowness Redux Props
 */
export interface AzimuthSlownessReduxProps {
  // passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  currentTimeInterval: CommonTypes.TimeRange;
  selectedSdIds: string[];
  openEventId: string;
  unassociatedSDColor: string;
  sdIdsToShowFk: string[];
  analysisMode: WorkflowTypes.AnalysisMode;
  location: AnalystWorkspaceTypes.LocationSolutionState;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  channelFilters: Record<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  setSelectedSdIds(ids: string[]): void;
  setChannelFilters(filters: Record<string, WaveformTypes.WaveformFilter>): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Record<string, boolean>): void;
}

export interface SubscriptionAction {
  (
    list: SignalDetectionTypes.SignalDetection[],
    index: number,
    prev: SignalDetectionTypes.SignalDetection[],
    currentIteree: SignalDetectionTypes.SignalDetection
  ): void;
}

/**
 * Azimuth Slowness State
 */
export interface AzimuthSlownessState {
  fkThumbnailSizePx: FkThumbnailSize;
  fkThumbnailColumnSizePx: number;
  filterType: FilterType;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  userInputFkFrequency: FkTypes.FrequencyBand;
  numberOfOutstandingComputeFkMutations: number;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  fkInnerContainerWidthPx: number;
  fkFrequencyThumbnails: Immutable.Map<string, FkTypes.FkFrequencyThumbnail[]>;
}

/**
 * Mutations used by the Az Slow display
 */
export interface AzimuthSlownessMutations {
  computeFks: (args: any) => Promise<void>;
  computeFkFrequencyThumbnails: (args: any) => Promise<void>;
  setWindowLead: (args: any) => Promise<void>;
  markFksReviewed: (args: any) => Promise<void>;
}

/**
 * Consolidated props for Azimuth Slowness
 */
export type AzimuthSlownessProps = AzimuthSlownessReduxProps &
  AzimuthSlownessMutations &
  ProcessingAnalystConfigurationQueryProps;

/**
 * State of the az slow panel
 */
export interface AzimuthSlownessPanelState {
  currentMovieSpectrumIndex: number;
}

export interface AzimuthSlownessPanelProps {
  // Data
  defaultStations: ProcessingStationTypes.ProcessingStation[];
  eventsInTimeRange: LegacyEventTypes.Event[];
  displayedSignalDetection: SignalDetectionTypes.SignalDetection | undefined;
  openEvent: LegacyEventTypes.Event | undefined;
  unassociatedSignalDetectionByColor: string;
  associatedSignalDetections: SignalDetectionTypes.SignalDetection[];
  signalDetectionsToDraw: SignalDetectionTypes.SignalDetection[];
  signalDetectionsIdToFeaturePredictions: Immutable.Map<
    string,
    LegacyEventTypes.FeaturePrediction[]
  >;
  signalDetectionsByStation: SignalDetectionTypes.SignalDetection[];
  featurePredictionsForDisplayedSignalDetection: LegacyEventTypes.FeaturePrediction[];
  distances: LegacyEventTypes.LocationToStationDistance[];
  // Redux state as props
  selectedSdIds: string[];
  defaultWaveformFilters: WaveformTypes.WaveformFilter[];
  sdIdsToShowFk: string[];
  location: AnalystWorkspaceTypes.LocationSolutionState;
  fkFrequencyThumbnails: FkTypes.FkFrequencyThumbnail[];
  // Azimuth display state as props
  fkThumbnailColumnSizePx: number;
  fkDisplayWidthPx: number;
  fkDisplayHeightPx: number;
  filterType: FilterType;
  fkThumbnailSizePx: FkThumbnailSize;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  numberOfOutstandingComputeFkMutations: number;
  userInputFkFrequency: FkTypes.FrequencyBand;
  fkUnitForDisplayedSignalDetection: FkUnits;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  fkInnerContainerWidthPx: number;
  channelFilters: Record<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  // Prop functions
  adjustFkInnerContainerWidth(
    fkThumbnailsContainer: HTMLDivElement,
    fkThumbnailsInnerContainer: HTMLDivElement
  ): void;
  markFksForSdIdsAsReviewed(sdIds: string[]): void;
  updateFkThumbnailSize(size: FkThumbnailSize): void;
  updateFkFilter(filterType: FilterType): void;
  setFkThumbnailColumnSizePx(newSizePx: number): void;
  computeFkAndUpdateState(fkInput: FkTypes.FkInput): void;
  changeUserInputFks(
    windowParams: FkTypes.WindowParameters,
    frequencyBand: FkTypes.FrequencyBand
  ): void;
  setFkUnitForSdId(sdId: string, fkUnit: FkUnits): void;

  // Redux functions
  setSelectedSdIds(sdIds: string[]): void;
  setSdIdsToShowFk(sdIds: string[]): void;
  setChannelFilters(filters: Record<string, WaveformTypes.WaveformFilter>): void;
  setMeasurementModeEntries(entries: Record<string, boolean>): void;
}
