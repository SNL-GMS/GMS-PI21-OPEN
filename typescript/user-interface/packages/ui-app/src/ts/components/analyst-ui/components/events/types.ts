import type { EventTypes } from '@gms/common-model';
import type { CellRendererParams, Row } from '@gms/ui-core-components';
import { EventsColumn } from '@gms/ui-state';
import Immutable from 'immutable';

/**
 * Event Table generic CellRendererParams type
 */
export type EventTableCellRendererParams<T> = CellRendererParams<
  EventRow,
  unknown,
  T,
  unknown,
  unknown
>;

/**
 * used to match the display strings to values in the site table column picker dropdown
 */
export const columnDisplayStrings: Immutable.Map<EventsColumn, string> = Immutable.Map<
  EventsColumn,
  string
>([
  [EventsColumn.conflict, 'Conflict'],
  [EventsColumn.time, 'Time'],
  [EventsColumn.latitudeDegrees, 'Lat (°)'],
  [EventsColumn.longitudeDegrees, 'Lon (°)'],
  [EventsColumn.depthKm, 'Depth (km)'],
  [EventsColumn.magnitudeMb, 'mb'],
  [EventsColumn.magnitudeMs, 'ms'],
  [EventsColumn.magnitudeMl, 'ml'],
  [EventsColumn.coverageSemiMajorAxis, 'Coverage semi-major'],
  [EventsColumn.coverageSemiMinorAxis, 'Coverage semi-minor'],
  [EventsColumn.confidenceSemiMajorAxis, 'Confidence semi-major'],
  [EventsColumn.confidenceSemiMinorAxis, 'Confidence semi-minor'],
  [EventsColumn.region, 'Region'],
  [EventsColumn.activeAnalysts, 'Active analysts'],
  [EventsColumn.preferred, 'Preferred in stage'],
  [EventsColumn.status, 'Workflow status'],
  [EventsColumn.rejected, 'Rejected']
]);

/**
 * Enumerated time types based on whether event or signal detection is within the interval or a before/after edge event
 */
export enum EdgeTypes {
  BEFORE = 'Before',
  AFTER = 'After',
  INTERVAL = 'Interval'
}

/**
 * Event row for event display
 */
export interface EventRow extends Row {
  readonly id: string;
  readonly edgeEventType: EdgeTypes;
  readonly isOpen: boolean;
  readonly conflict: boolean;
  readonly time: number;
  readonly latitudeDegrees: number;
  readonly longitudeDegrees: number;
  readonly depthKm: number;
  readonly region: string;
  readonly confidenceSemiMajorAxis: number;
  readonly confidenceSemiMinorAxis: number;
  readonly coverageSemiMajorAxis: number;
  readonly coverageSemiMinorAxis: number;
  readonly magnitudeMb: number;
  readonly magnitudeMs: number;
  readonly magnitudeMl: number;
  readonly activeAnalysts: string[];
  readonly preferred: string;
  readonly status: EventTypes.EventStatus;
  readonly rejected: string;
}
