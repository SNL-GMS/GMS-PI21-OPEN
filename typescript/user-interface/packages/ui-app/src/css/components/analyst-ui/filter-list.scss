.filters-display-window {
  height: 100%;
  flex-grow: 1;
}

@mixin listReset {
  list-style: none;
  margin-block-start: 0;
  margin-block-end: 0;
  margin-inline-start: 0;
  margin-inline-end: 0;
  padding-inline-start: 0;
}

@mixin selectedAndActive {
  background-color: var(--gms-selection, $gms-selection);
  filter: brightness(1.2);
  color: var(--gms-main, $gms-main);
}

@mixin listEntry($selectable: true) {
  background-color: var(--gms-prominent-background, $gms-prominent-background);
  border-radius: 3px;
  cursor: pointer;
  padding: 0.5em;
  transition: background-color 0.2s ease;
  @if ($selectable) {
    &--selected,
    &--selected.bp4-selected.bp4-menu-item,
    .filter-list-entry--selected & {
      background-color: var(--gms-selection, $gms-selection);
      color: var(--gms-main, $gms-main);
      &:hover,
      &:active,
      &.bp4-active,
      &.bp4-active.bp4-menu-item {
        @include selectedAndActive;
      }
    }
    .filter-list-entry--selected.filter-list-entry--active & {
      @include selectedAndActive;
    }
    .filter-list-entry--active & {
      background-color: rgba(45, 114, 210, 0.2); // light mode: rgba(45, 114, 210, 0.1);
      color: var(--active-text-color, $gms-main); // light mode: #215db0
    }
  }
  &:hover {
    background-color: var(--gms-button, $gms-button);
  }
}

@mixin listRow($lineHeight: 1) {
  line-height: $lineHeight;
  margin: 2px 0;
}

.filter-list-picker {
  &__container {
    @include listReset;
    width: 100%;
  }
  &__option {
    @include listEntry;
    @include listRow(auto);
  }
}

.filter-list {
  @include listReset;
  &__container {
    &--scrollable {
      position: relative;
      margin-top: 0.25em;
      flex-grow: 1;
      overflow-y: auto;
      display: grid;
      grid-template-rows: auto 1fr;
    }
    & &__header {
      font-size: 1.125rem;
    }
  }
  &__button {
    width: 100%;
    justify-content: flex-start;
  }
}

.filter-list-entry {
  @include listRow;
  display: flex;
  gap: 2px;
  &__icon {
    @include listEntry(false);
    background-color: transparent;
    padding: 0;
  }
  &__button {
    @include listEntry(false);
    align-items: center;
    // Compatible with target browsers
    aspect-ratio: 1 / 1;
    display: flex;
    background-color: transparent;
    &--hidden {
      opacity: 0;
      &:hover {
        opacity: 1;
      }
    }
  }
  &__info {
    right: 0;
    margin-left: auto;
    opacity: 0;
    padding: 0.5em;
    position: sticky;
    top: 0;
    transition: opacity 0.3s ease;
    // Negative margins allow us to use sticky positioning
    // and still make our info button be the same size as the row.
    margin-bottom: -0.5em;
    margin-right: -2em;
    margin-top: -0.5em;
    background-color: inherit;
    & .bp4-icon {
      opacity: 0;
    }
    .filter-list-entry:hover & {
      opacity: 1;
      &:hover {
        & .bp4-icon {
          opacity: 1;
        }
      }
      & .bp4-icon {
        opacity: 0.5;
      }
    }
  }
  &__title {
    @include listEntry;
    display: flex;
    flex-grow: 1;
    padding-right: 2em;
    position: relative;
    white-space: nowrap;
  }
  &__tooltip-wrapper {
    flex-grow: 1;
  }
}

.filter-tooltip-content {
  .bp4-heading {
    margin-top: 0;
    margin-left: 0;
    font-weight: bold;
    &:last-child {
      // in case there is no content below
      margin-bottom: 0;
      margin-top: 0;
    }
  }
}

.filter-list-tooltip {
  &__container {
    display: grid;
    background-color: rgba(255, 255, 255, 50%);
    padding: 0 0.5em;
    grid-template-columns: 16ch auto;
    & .label-value__label {
      font-weight: 500;
      margin-right: 1em;
    }
  }
  &__group {
    gap: 2px;
    display: flex;
    flex-direction: column;
    & .label-value__label {
      padding-left: 1em;
      position: relative;

      /// Make a circular bullet to the left of the indented label using after pseudoelement
      &::after {
        content: '';
        position: absolute;
        display: block;
        line-height: 0.25em;
        top: 50%;
        left: 0.25em;
        transform: translate3D(0, -50%, 0);
        border-radius: 1rem;
        width: 0.25rem;
        aspect-ratio: 1/1;
        background-color: var(--gms-main-inverted, $gms-main-inverted);
      }
    }
  }
}

.filter-tooltip-content {
  display: flex;
  flex-direction: column;
  max-width: 60ch;
  gap: 2px;
  font-size: 0.875em;
}
