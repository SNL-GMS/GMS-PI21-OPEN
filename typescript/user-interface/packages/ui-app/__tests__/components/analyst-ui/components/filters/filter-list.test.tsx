/* eslint-disable @typescript-eslint/no-magic-numbers */
import { BandType, FilterType } from '@gms/common-model/lib/filter-list/types';
import { analystActions, getStore } from '@gms/ui-state';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import {
  buildHandleKeyDown,
  FilterListOrNonIdealState
} from '../../../../../src/ts/components/analyst-ui/components/filters/filter-list';

const mockSetSelectedFilter = jest.fn();
const mockSetSelectedFilterIndex = jest.fn();
jest.mock('@gms/ui-state', () => {
  const actual = jest.requireActual('@gms/ui-state');
  return {
    ...actual,
    useUiTheme: () => [
      {
        colors: {
          gmsSelection: '#123123',
          gmsMain: '#BADBAD',
          gmsMainInverted: '#DABDAB',
          gmsBackground: '#000000'
        }
      },
      jest.fn()
    ],
    useSelectedFilter: () => ({
      selectedFilterList: {
        withinHotKeyCycle: true,
        unfiltered: true,
        namedFilter: null,
        filterDefinition: null
      },
      setSelectedFilter: mockSetSelectedFilter
    }),
    useSelectedFilterIndex: () => ({
      selectedFilterIndex: 0,
      setSelectedFilterIndex: mockSetSelectedFilterIndex
    }),
    useHotkeyCycle: () => ({
      hotkeyCycle: [true, false, true],
      setIsFilterWithinHotkeyCycle: jest.fn()
    })
  };
});

describe('FilterList', () => {
  const store = getStore();
  store.dispatch(analystActions.setSelectedFilterIndex(0));
  store.dispatch(analystActions.setSelectedFilterList('test'));
  store.dispatch(
    analystActions.setFilterHotkeyCycleOverridesForCurrentFilterList({
      0: false,
      1: true,
      2: true
    })
  );

  const nonIdeal = renderer.create(
    <Provider store={store}>
      <FilterListOrNonIdealState filterList={undefined} />
    </Provider>
  );
  it('matches a non ideal state snapshot when given no filter list', () => {
    expect(nonIdeal.toJSON()).toMatchSnapshot();
  });
  const tree = renderer.create(
    <Provider store={store}>
      <FilterListOrNonIdealState
        filterList={{
          name: 'test',
          defaultFilterIndex: 0,
          filters: [
            {
              filterDefinition: {
                comments: 'the comments 1',
                filterDescription: {
                  causal: false,
                  comments: 'the description comments 1',
                  filterType: FilterType.IIR_BUTTERWORTH,
                  highFrequency: 1,
                  lowFrequency: 0.5,
                  order: 1,
                  parameters: {
                    aCoefficients: [0.1, 1.0],
                    bCoefficients: [1.1, 1.2],
                    groupDelaySec: 3,
                    sampleRateHz: 40,
                    sampleRateToleranceHz: 2
                  },
                  passBandType: BandType.BAND_PASS,
                  zeroPhase: false
                },
                name: 'filter def name 1'
              },
              namedFilter: 'test filter 1',
              unfiltered: false,
              withinHotKeyCycle: true
            },
            {
              filterDefinition: {
                comments: 'the comments 2',
                filterDescription: {
                  causal: true,
                  comments: 'the description comments 2',
                  filterType: FilterType.IIR_BUTTERWORTH,
                  highFrequency: 2,
                  lowFrequency: 0.25,
                  order: 1,
                  parameters: {
                    aCoefficients: [0.2, 2.0],
                    bCoefficients: [2, 2.2],
                    groupDelaySec: 2,
                    sampleRateHz: 20,
                    sampleRateToleranceHz: 22
                  },
                  passBandType: BandType.BAND_PASS,
                  zeroPhase: true
                },
                name: 'filter def name 1'
              },
              namedFilter: 'test filter 1',
              unfiltered: false,
              withinHotKeyCycle: false
            },
            {
              filterDefinition: undefined,
              namedFilter: undefined,
              unfiltered: true,
              withinHotKeyCycle: true
            }
          ]
        }}
      />
    </Provider>
  );
  it('matches a snapshot', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });
  describe('buildHandleKeyDown', () => {
    const mockSetActiveIndex = jest.fn();
    const mockScrollIntoView = jest.fn();
    const mockHandleBlur = jest.fn();
    const mockSetSelectedFilterIndex2 = jest.fn(); // 2 because we have one with the same name above
    const mockFilterList = { filters: [{}, {}, {}] };
    const testHandleKeyDown = buildHandleKeyDown(
      mockSetActiveIndex,
      mockScrollIntoView,
      mockHandleBlur,
      mockSetSelectedFilterIndex2,
      mockFilterList,
      0
    );
    beforeEach(() => {
      mockSetActiveIndex.mockClear();
      mockScrollIntoView.mockClear();
      mockHandleBlur.mockClear();
      mockSetSelectedFilterIndex2.mockClear();
    });
    test('calls handleBlur when the escape key is pressed', () => {
      testHandleKeyDown({ key: 'Escape', preventDefault: jest.fn(), stopPropagation: jest.fn() });
      expect(mockHandleBlur).toHaveBeenCalled();
    });
    test('sets the selected active filter index when enter is pressed', () => {
      testHandleKeyDown({ key: 'Enter', preventDefault: jest.fn(), stopPropagation: jest.fn() });
      expect(mockSetSelectedFilterIndex2).toHaveBeenCalledWith(0);
    });
    test('sets the previous index when ArrowUp key is pressed', () => {
      testHandleKeyDown({ key: 'ArrowUp', preventDefault: jest.fn(), stopPropagation: jest.fn() });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      expect(setterFn(2)).toBe(1);
      expect(setterFn(1)).toBe(0);
      expect(setterFn(0)).toBe(2);
    });
    test('calls scrollIntoView with the previous index when ArrowUp key is pressed', () => {
      testHandleKeyDown({ key: 'ArrowUp', preventDefault: jest.fn(), stopPropagation: jest.fn() });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      setterFn(2);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(1);
      setterFn(1);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(0);
      setterFn(0);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(2);
    });
    test('sets the previous index when Shift + Tab is pressed', () => {
      testHandleKeyDown({
        key: 'Tab',
        shiftKey: true,
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      expect(setterFn(2)).toBe(1);
      expect(setterFn(1)).toBe(0);
      expect(setterFn(0)).toBe(2);
    });
    test('sets the next index when ArrowDown key is pressed', () => {
      testHandleKeyDown({
        key: 'ArrowDown',
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      expect(setterFn(2)).toBe(0);
      expect(setterFn(1)).toBe(2);
      expect(setterFn(0)).toBe(1);
    });
    test('calls scrollIntoView when the next index when ArrowDown key is pressed', () => {
      testHandleKeyDown({
        key: 'ArrowDown',
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      setterFn(2);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(0);
      setterFn(1);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(2);
      setterFn(0);
      expect(mockScrollIntoView).toHaveBeenLastCalledWith(1);
    });
    test('sets the next index when Tab key is pressed', () => {
      testHandleKeyDown({
        key: 'Tab',
        shiftKey: false,
        preventDefault: jest.fn(),
        stopPropagation: jest.fn()
      });
      const setterFn = mockSetActiveIndex.mock.calls[mockSetActiveIndex.mock.calls.length - 1][0];
      expect(typeof setterFn).toBe('function');
      expect(setterFn(2)).toBe(0);
      expect(setterFn(1)).toBe(2);
      expect(setterFn(0)).toBe(1);
    });
  });
});
