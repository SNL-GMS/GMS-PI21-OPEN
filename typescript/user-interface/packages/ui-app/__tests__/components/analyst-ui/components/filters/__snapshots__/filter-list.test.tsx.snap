// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterList matches a non ideal state snapshot when given no filter list 1`] = `
<div
  className="bp4-non-ideal-state"
>
  <h4
    className="bp4-heading"
  >
    Filter List
  </h4>
  <div>
    No filter lists found
  </div>
</div>
`;

exports[`FilterList matches a snapshot 1`] = `
<section
  className="filter-list__container filter-list__container--scrollable"
  onBlur={[Function]}
  onFocusCapture={[Function]}
  onKeyDown={[Function]}
  role="menu"
  style={
    Object {
      "--active-text-color": "#668668",
    }
  }
  tabIndex={-1}
>
  <h6
    className="bp4-heading filter-list__header"
  >
    Filters
  </h6>
  <ol
    className="filter-list"
  >
    <li
      className="filter-list-entry filter-list-entry--selected"
      onClick={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        aria-hidden={true}
        className="bp4-icon bp4-icon-star filter-list-entry__icon filter-list-entry__button"
        icon="star"
        onClick={[Function]}
      >
        <svg
          data-icon="star"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M8 0l2.5 5.3 5.5.8-4 4.1.9 5.8L8 13.3 3.1 16l.9-5.8-4-4.1 5.5-.8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        aria-haspopup="true"
        className="filter-list-entry__tooltip-wrapper bp4-popover2-target"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="filter-list-entry__title"
          tabIndex={0}
        >
          filter def name 1
          <span
            aria-haspopup="true"
            className="filter-list-entry__info filter-list-entry__icon bp4-popover2-target"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              aria-hidden={true}
              className="bp4-icon bp4-icon-info-sign"
              icon="info-sign"
              tabIndex={0}
            >
              <svg
                data-icon="info-sign"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <path
                  d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </span>
        </div>
      </span>
    </li>
    <li
      className="filter-list-entry"
      onClick={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        aria-hidden={true}
        className="bp4-icon bp4-icon-star-empty filter-list-entry__icon filter-list-entry__button filter-list-entry__button--hidden"
        icon="star-empty"
        onClick={[Function]}
      >
        <svg
          data-icon="star-empty"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M16 6.11l-5.53-.84L8 0 5.53 5.27 0 6.11l4 4.1L3.06 16 8 13.27 12.94 16 12 10.21l4-4.1zM4.91 13.2l.59-3.62L3 7.02l3.45-.53L8 3.2l1.55 3.29 3.45.53-2.5 2.56.59 3.62L8 11.49 4.91 13.2z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        aria-haspopup="true"
        className="filter-list-entry__tooltip-wrapper bp4-popover2-target"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="filter-list-entry__title"
          tabIndex={0}
        >
          filter def name 1
          <span
            aria-haspopup="true"
            className="filter-list-entry__info filter-list-entry__icon bp4-popover2-target"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              aria-hidden={true}
              className="bp4-icon bp4-icon-info-sign"
              icon="info-sign"
              tabIndex={0}
            >
              <svg
                data-icon="info-sign"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <path
                  d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </span>
        </div>
      </span>
    </li>
    <li
      className="filter-list-entry"
      onClick={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        aria-hidden={true}
        className="bp4-icon bp4-icon-star filter-list-entry__icon filter-list-entry__button"
        icon="star"
        onClick={[Function]}
      >
        <svg
          data-icon="star"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M8 0l2.5 5.3 5.5.8-4 4.1.9 5.8L8 13.3 3.1 16l.9-5.8-4-4.1 5.5-.8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        aria-haspopup="true"
        className="filter-list-entry__tooltip-wrapper bp4-popover2-target"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="filter-list-entry__title"
          tabIndex={0}
        >
          Unfiltered
          <span
            aria-haspopup="true"
            className="filter-list-entry__info filter-list-entry__icon bp4-popover2-target"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              aria-hidden={true}
              className="bp4-icon bp4-icon-info-sign"
              icon="info-sign"
              tabIndex={0}
            >
              <svg
                data-icon="info-sign"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <path
                  d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </span>
        </div>
      </span>
    </li>
  </ol>
</section>
`;
